dawn/test/unit-test/dawn/Optimizer/Passes/TestTemporaryToFunction.cpp

```diff
@@ -12,54 +12,95 @@
 //
 //===------------------------------------------------------------------------------------------===//
 
-#include "dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.h"
-#include "dawn/CodeGen/GridTools/GTCodeGen.h"
 #include "dawn/Compiler/DawnCompiler.h"
-#include "dawn/Serialization/SIRSerializer.h"
+#include "dawn/Compiler/Options.h"
+#include "dawn/IIR/IIR.h"
+#include "dawn/IIR/LoopOrder.h"
+#include "dawn/IIR/MultiStage.h"
+#include "dawn/IIR/Stage.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Optimizer/PassMultiStageSplitter.h"
+#include "dawn/Optimizer/PassTemporaryToStencilFunction.h"
+#include "dawn/Serialization/IIRSerializer.h"
+#include "dawn/Unittest/CompilerUtil.h"
 #include "test/unit-test/dawn/Optimizer/TestEnvironment.h"
+
 #include <fstream>
 #include <gtest/gtest.h>
 
 using namespace dawn;
 
 namespace {
 
-class TemporaryToFunction : public ::testing::Test {
-
-  dawn::DawnCompiler compiler_;
+class TestPassTemporaryToFunction : public ::testing::Test {
+public:
+  TestPassTemporaryToFunction() {
+    options_.TmpToStencilFunction = true;
+    context_ = std::make_unique<OptimizerContext>(diagnostics_, options_, nullptr);
+  }
 
 protected:
-  TemporaryToFunction() { compiler_.getOptions().TmpToStencilFunction = true; }
-  virtual void SetUp() {}
-
-  std::shared_ptr<iir::StencilInstantiation> loadTest(std::string sirFilename) {
-
-    std::string filename = TestEnvironment::path_ + "/" + sirFilename;
-    std::ifstream file(filename);
-    DAWN_ASSERT_MSG((file.good()), std::string("File " + filename + " does not exists").c_str());
-
-    std::string jsonstr((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+  dawn::OptimizerContext::OptimizerContextOptions options_;
+  DiagnosticsEngine diagnostics_;
+  std::unique_ptr<OptimizerContext> context_;
 
-    std::shared_ptr<SIR> sir =
-        SIRSerializer::deserializeFromString(jsonstr, SIRSerializer::Format::Json);
+  std::shared_ptr<iir::StencilInstantiation> runPass(const std::string& filename) {
+    context_->getDiagnostics().clear();
+    std::shared_ptr<iir::StencilInstantiation> instantiation = IIRSerializer::deserialize(filename);
+    EXPECT_TRUE(instantiation->getIIR()->getChildren().size() == 1);
+    const auto& stencil = *instantiation->getIIR()->getChildren().front();
 
-    auto stencilInstantiationMap = compiler_.optimize(compiler_.lowerToIIR(sir));
-    // Report diagnostics
-    if(compiler_.getDiagnostics().hasDiags()) {
-      for(const auto& diag : compiler_.getDiagnostics().getQueue())
-        std::cerr << "Compilation Error " << diag->getMessage() << std::endl;
-      throw std::runtime_error("compilation failed");
+    // temp to function pass expects stmt data to contain the stack trace (can be empty though)
+    for(const auto& multiStage : stencil.getChildren()) {
```

Comment:
can we collapse this into one loop?

Reply:
I'm not sure and its not the first time I have been wondering how to achieve this using our code base. I could obviously do it by stringing together this Frankensteinian construct:
```
for(const auto& stmt : stencil.getChildren()
                               .front()
                               ->getChildren()
                               .front()
                               ->getChildren()
                               .front()
                               ->getAST()
                               .getStatements()) {
      stmt->getData<iir::IIRStmtData>().StackTrace = std::vector<dawn::ast::StencilCall*>();
    }
```
But is this really what you/we want? Maybe a utility class blindly stringing together all statements in a stencilInstantiation is in order?

Reply:
that is why we have IterateIIROver